pipeline/ruleset.go

```diff
@@ -122,170 +125,170 @@ func (r *Rules) Empty() bool {
 // ruletypes from the rules match the provided ruledata. For
 // both operators, when none of the ruletypes from the rules
 // match the provided ruledata, the function returns false.
-//
-//nolint:gocyclo // accepting complexity in this case
-func (r *Rules) Match(from *RuleData, matcher, op string) bool {
-	// set defaults
-	status := true
+func (r *Rules) Match(from *RuleData, matcher, op string) (bool, error) {
+	var err error
 
 	// if the path ruletype is provided
 	if len(from.Path) > 0 {
 		// if the "or" operator is provided in the ruleset
 		if strings.EqualFold(op, constants.OperatorOr) {
-			// override the default to the "or"
-			if len(from.Status) != 0 {
-				status = r.Status.MatchOr(from.Status, matcher)
-			}
-
 			// iterate through each path in the ruletype
 			for _, p := range from.Path {
+				matches, err := matches(r, from, matcher, p, constants.OperatorOr)
+				if err != nil {
+					return false, err
+				}
+
 				// return true if any ruletype matches the ruledata
-				if r.Branch.MatchOr(from.Branch, matcher) ||
-					r.Comment.MatchOr(from.Comment, matcher) ||
-					r.Event.MatchOr(from.Event, matcher) ||
-					r.Path.MatchOr(p, matcher) ||
-					r.Repo.MatchOr(from.Repo, matcher) ||
-					status ||
-					r.Tag.MatchOr(from.Tag, matcher) ||
-					r.Target.MatchOr(from.Target, matcher) {
-					return true
+				if matches {
+					return true, nil
 				}
 			}
 
 			// return false if no match is found
-			return false
-		}
-
-		// override the default to the "and"
-		if len(from.Status) != 0 {
-			status = r.Status.MatchAnd(from.Status, matcher)
+			return false, nil
 		}
 
 		// iterate through each path in the ruletype
 		for _, p := range from.Path {
-			// return true if every ruletype matches the ruledata
-			if r.Branch.MatchAnd(from.Branch, matcher) &&
-				r.Comment.MatchAnd(from.Comment, matcher) &&
-				r.Event.MatchAnd(from.Event, matcher) &&
-				r.Path.MatchAnd(p, matcher) &&
-				r.Repo.MatchAnd(from.Repo, matcher) &&
-				status &&
-				r.Tag.MatchAnd(from.Tag, matcher) &&
-				r.Target.MatchAnd(from.Target, matcher) {
-				return true
+			matches, err := matches(r, from, matcher, p, constants.OperatorAnd)
+			if err != nil {
+				return false, err
+			}
+
+			// return true if any ruletype matches the ruledata
+			if matches {
+				return true, nil
 			}
```

Comment:
could do `return matches(r, from, matcher, p, constants.OperatorAnd)`

Reply:
We want to continue looping through the paths if there's no match, so returning `false, nil` is not an option here unfortunately...

Reply:
hmm, maybe we need to adjust tests since i didn't see any test fails with the change?!